pkg/apiserver/statement/statement.go

```diff
@@ -291,24 +294,42 @@ func (s *Service) downloadTokenHandler(c *gin.Context) {
 		csvData = append(csvData, row)
 	}
 
-	// generate csv
+	// generate temp file that persist encrypted data
 	timeLayout := "01021504"
 	beginTime := time.Unix(int64(req.BeginTime), 0).Format(timeLayout)
 	endTime := time.Unix(int64(req.EndTime), 0).Format(timeLayout)
-	tmpfile, err := ioutil.TempFile("", fmt.Sprintf("statements_%s_%s_*.csv", beginTime, endTime))
+	csvFile, err := ioutil.TempFile("", fmt.Sprintf("statements_%s_%s_*.csv", beginTime, endTime))
 	if err != nil {
 		_ = c.Error(err)
 		return
 	}
-	defer tmpfile.Close()
-	csvwriter := csv.NewWriter(tmpfile)
+	defer csvFile.Close()
+	// generate buf writer for tmpfile
+	csvBuf := bufio.NewWriter(csvFile)
+	// generate encrypted key
+	secretKey := cryptopasta.NewEncryptionKey()
+
+	rowBuf := bytes.NewBuffer(nil)
+	csvwriter := csv.NewWriter(rowBuf)
```

Comment:
This library implements an encrypted io.Reader / io.Writer, which could be a better choice. Maybe you can have a try: https://github.com/Xeoncross/go-aesctr-with-hmac

The encapsulation of the stream is file stream -> encrypted stream -> csv stream.

Reply:
Ok, thank you! let me have a look. Since I found zip the encrypted file won't reduce the size much, so I think it doesn't need to zip file in local, right? but to still zip it when sending data back to the client after being decrypted.

Reply:
IIRC, we use gzip already. Is there any further benefit of packaging as a zip for transmission?

Reply:
Make sense, let me focus on the encryption first.

Reply:
It works fine! and the final flow is this:

![image](https://user-images.githubusercontent.com/1284531/97527773-14151d80-19e7-11eb-9c36-6c868f65e0f8.png)